src/context/LocationsContext.js

```diff
@@ -1,61 +1,79 @@
 import React, { createContext, useEffect, useState } from 'react';
+import { getLocations } from '../lib/fetchPlaces';
 
 export const LocationsContext = createContext();
 
 const LocationsContextProvider = (props) => {
   const [locations, setLocations] = useState([]);
+  const [userLocation, setUserLocation] = useState();
+  const [allowLocation, setAllowLocation] = useState(false);
   //state is updated in the Map component
   const [newCenter, setNewCenter] = useState();
+  const [loading, setLoading] = useState({ loading: false, message: '' });
 
   //New Orleans
   const defaultCoordinates = {
     lat: 29.9511,
     lng: -90.0715,
   };
 
-  //leaving room for the logic from the other groups ticket
-  const coordinates = defaultCoordinates;
+  const userCoordinates = userLocation && {
+    lat: userLocation.latitude,
+    lng: userLocation.longitude,
+  };
 
-  const url = `https://segdeha.com/api/nearby.php?lat=${coordinates.lat}&lng=${coordinates.lng}`;
+  let coordinates = userLocation ? userCoordinates : defaultCoordinates;
 
   useEffect(() => {
-    fetch(url)
-      .then((response) => {
-        return response.json();
-      })
-      .then((response) => {
-        let pages = response.query ? response.query.pages : [];
-        setLocations(pages);
-      })
-      .catch(console.log);
-  }, []);
+    getLocations(coordinates.lat, coordinates.lng, setLocations);
+
+    if (navigator.geolocation && allowLocation) {
+      setLoading({ loading: true, message: 'loading' });
+      navigator.geolocation.getCurrentPosition(
+        (position) => {
+          setUserLocation(position.coords);
+          setLoading({ loading: false, message: '' });
+        },
+        (error) => {
+          console.error(error);
+          setLoading({
+            loading: false,
+            message: 'location services turned off',
+          });
+        },
+      );
+    }
+  }, [allowLocation]);
 
   /*when the newCenter changes in the map componentt the useEffect
   makes a new api call & the new locations are updated */
   useEffect(() => {
-    let lat, lng;
     if (newCenter) {
-      lat = newCenter.lat;
-      lng = newCenter.lng;
-    } else {
-      lat = defaultCoordinates.lat;
-      lng = defaultCoordinates.lng;
+      getLocations(newCenter.lat, newCenter.lng, setLocations);
     }
-    const newCenterUrl = `https://segdeha.com/api/nearby.php?lat=${lat}&lng=${lng}`;
-    fetch(newCenterUrl)
-      .then((response) => {
-        return response.json();
-      })
-      .then((response) => {
-        let pages = response.query ? response.query.pages : [];
-        setLocations(pages);
-      })
-      .catch(console.log);
   }, [newCenter]);
 
+  // When allowLocation is false clear state of user location and reset center to map center
+  useEffect(() => {
```

Comment:
So, there are still 3 different `useEffect`s that can potentially call `getLocations` ... are we sure they're all mutually exclusive? In other words, will there ever be a case where more than one thinks it should be the one fetching the new set of places?

Reply:
What if we consolidate to this?

 /*when the newCenter changes in the map componentt the useEffect
  makes a new api call & the new locations are updated 
  when userLocation is set, the new locations are updated!*/
  
`useEffect(() => {
    if (newCenter) {
      getLocations(newCenter.lat, newCenter.lng, setLocations);
    }
    if(userLocation){
      getLocations(coordinates.lat, coordinates.lng, setLocations);
    }
  }, [newCenter, userLocation]);`

It works for me! @segdeha @snsebro

Reply:
> When the `newCenter` changes in the map component the `useEffect` makes a new API call & the new locations are updated when `userLocation` is set, the new locations are updated!

Yeah, that sounds right! Whatever makes the API call should be controlled by a single input. Good thinking, @rachel-fischoff!